### PR TITLE
Add team contact resource

### PR DESCRIPTION
**Pull request by taimoorgit**
## Issues

https://github.com/OpsLevel/team-platform/issues/308

https://registry.terraform.io/providers/OpsLevel/opslevel/latest/docs/resources/team_contact

## Changelog

- [x] Add team contact

## Tophatting

Config file:

```
resource "opslevel_team_contact" "tc_1" {
  name = "Slack Channel"
  team = "team_platform_3"
  type = "slack"
  value = "#team-platform-3"
}

resource "opslevel_team_contact" "tc_2" {
  name = "Internal Email Address"
  team = "team_platform_3"
  type = "email"
  value = "team-platform-3@opslevel.com"
}

```

### Create

```
Terraform will perform the following actions:

  # opslevel_team_contact.tc_1 will be created
  + resource "opslevel_team_contact" "tc_1" {
      + id           = (known after apply)
      + last_updated = (known after apply)
      + name         = "Slack Channel"
      + team         = "team_platform_3"
      + type         = "slack"
      + value        = "#team-platform-3"
    }

  # opslevel_team_contact.tc_2 will be created
  + resource "opslevel_team_contact" "tc_2" {
      + id           = (known after apply)
      + last_updated = (known after apply)
      + name         = "Internal Email Address"
      + team         = "team_platform_3"
      + type         = "email"
      + value        = "team-platform-3@opslevel.com"
    }

Plan: 2 to add, 0 to change, 0 to destroy.
opslevel_team_contact.tc_2: Creating...
opslevel_team_contact.tc_1: Creating...
opslevel_team_contact.tc_1: Creation complete after 0s [id=Z2lkOi8vb3BzbGV2ZWwvQ29udGFjdC8zODc1Nw]
opslevel_team_contact.tc_2: Creation complete after 0s [id=Z2lkOi8vb3BzbGV2ZWwvQ29udGFjdC8zODc1OA]

Apply complete! Resources: 2 added, 0 changed, 0 destroyed.

```

image

### Update

```
Terraform will perform the following actions:

  # opslevel_team_contact.tc_1 will be updated in-place
  ~ resource "opslevel_team_contact" "tc_1" {
        id           = "Z2lkOi8vb3BzbGV2ZWwvQ29udGFjdC8zODc1Nw"
      + last_updated = (known after apply)
      ~ name         = "Slack Channel" -> "Internal Slack Channel"
      ~ value        = "#team-platform-3" -> "#platform-3"
        # (2 unchanged attributes hidden)
    }

  # opslevel_team_contact.tc_2 will be updated in-place
  ~ resource "opslevel_team_contact" "tc_2" {
        id           = "Z2lkOi8vb3BzbGV2ZWwvQ29udGFjdC8zODc1OA"
      + last_updated = (known after apply)
      ~ name         = "Internal Email Address" -> "Team Email Internal"
      ~ value        = "team-platform-3@opslevel.com" -> "team-platform-3-3-3@opslevel.com"
        # (2 unchanged attributes hidden)
    }

Plan: 0 to add, 2 to change, 0 to destroy.
opslevel_team_contact.tc_2: Modifying... [id=Z2lkOi8vb3BzbGV2ZWwvQ29udGFjdC8zODc1OA]
opslevel_team_contact.tc_1: Modifying... [id=Z2lkOi8vb3BzbGV2ZWwvQ29udGFjdC8zODc1Nw]
opslevel_team_contact.tc_2: Modifications complete after 0s [id=Z2lkOi8vb3BzbGV2ZWwvQ29udGFjdC8zODc1OA]
opslevel_team_contact.tc_1: Modifications complete after 0s [id=Z2lkOi8vb3BzbGV2ZWwvQ29udGFjdC8zODc1Nw]

Apply complete! Resources: 0 added, 2 changed, 0 destroyed.

```

image

### Delete

```
Terraform will perform the following actions:

  # opslevel_team_contact.tc_1 will be destroyed
  - resource "opslevel_team_contact" "tc_1" {
      - id    = "Z2lkOi8vb3BzbGV2ZWwvQ29udGFjdC8zODc1Nw" -> null
      - name  = "Internal Slack Channel" -> null
      - team  = "team_platform_3" -> null
      - type  = "slack" -> null
      - value = "#platform-3" -> null
    }

  # opslevel_team_contact.tc_2 will be destroyed
  - resource "opslevel_team_contact" "tc_2" {
      - id    = "Z2lkOi8vb3BzbGV2ZWwvQ29udGFjdC8zODc1OA" -> null
      - name  = "Team Email Internal" -> null
      - team  = "team_platform_3" -> null
      - type  = "email" -> null
      - value = "team-platform-3-3-3@opslevel.com" -> null
    }

Plan: 0 to add, 0 to change, 2 to destroy.
opslevel_team_contact.tc_1: Destroying... [id=Z2lkOi8vb3BzbGV2ZWwvQ29udGFjdC8zODc1Nw]
opslevel_team_contact.tc_2: Destroying... [id=Z2lkOi8vb3BzbGV2ZWwvQ29udGFjdC8zODc1OA]
opslevel_team_contact.tc_1: Destruction complete after 1s
opslevel_team_contact.tc_2: Destruction complete after 1s

Destroy complete! Resources: 2 destroyed.

```

### CODE REVIEW

File: opslevel/resource_opslevel_team_contact.go [hunk 1]
```diff
@@ -1,146 +1,204 @@
...
+	Team        types.String `tfsdk:"team"`
+	Type        types.String `tfsdk:"type"`
+	Value       types.String `tfsdk:"value"`
+}
+
+func NewTeamContactResourceModel(teamContact opslevel.Contact, teamIdentifier string) TeamContactResourceModel {
+	teamResourceModel := TeamContactResourceModel{
+		Id:    types.StringValue(string(teamContact.Id)),
+		Name:  RequiredStringValue(teamContact.DisplayName),
+		Team:  types.StringValue(teamIdentifier),
```
Comment (taimoorgit) [inline opslevel/resource_opslevel_team_contact.go:42]: TODO: use type conversion functions

File: opslevel/resource_opslevel_team_contact.go [hunk 2]
```diff
@@ -1,146 +1,204 @@
...
+		Attributes: map[string]schema.Attribute{
+			"id": schema.StringAttribute{
+				Description: "The ID of this resource.",
+				Computed:    true,
+				PlanModifiers: []planmodifier.String{
+					stringplanmodifier.UseStateForUnknown(),
+				},
+			},
+			"last_updated": schema.StringAttribute{
+				Optional: true,
```
Comment (taimoorgit) [inline opslevel/resource_opslevel_team_contact.go:65]: TODO: remove optional here

File: opslevel/resource_opslevel_team_contact.go [hunk 3]
```diff
@@ -1,146 +1,204 @@
...
+				Description: "The name shown in the UI for the contact.",
+				Required:    true,
+			},
+			"team": schema.StringAttribute{
+				Description: "The id or alias of the team the contact belongs to.",
+				Required:    true,
+			},
+			"type": schema.StringAttribute{
+				Description: "The method of contact [email, slack, slack_handle, web].",
+				Required:    true,
```
Comment (davidbloss) [inline opslevel/resource_opslevel_team_contact.go:78]: TODO: add validator

File: opslevel/resource_opslevel_team_contact.go [hunk 4]
```diff
@@ -1,146 +1,211 @@
...
+func (teamContactResource *TeamContactResource) Update(ctx context.Context, req resource.UpdateRequest, resp *resource.UpdateResponse) {
+	var data TeamContactResourceModel
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &data)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	contactCreateInput := opslevel.ContactInput{
+		Address:     data.Value.ValueString(),
+		DisplayName: data.Name.ValueStringPointer(),
```
Comment (davidbloss) [inline opslevel/resource_opslevel_team_contact.go:172]:
```suggestion
		DisplayName: opslevel.RefOf(data.Name.ValueString()),
```
I found this subtle bug today. For Update only, this change in code will support unsetting this optional field

Comment (davidbloss) [reply opslevel/resource_opslevel_team_contact.go:172]: `data.Name.ValueStringPointer(),` will return `nil` and opslevel-go will ignore this field. But we actually want to unset it here so we need to provide a string pointer to a blank string

**Review by davidbloss — APPROVED** [review]
I left one comment on a needed fix. But once that's in this can be merged 🎉

File: opslevel/resource_opslevel_team_contact.go (hunk 4 above)
Comment (taimoorgit) [reply opslevel/resource_opslevel_team_contact.go]: @davidbloss done

**Pull request merged**